### PR TITLE
Add toView . fromView rule

### PR DESCRIPTION
**Pull request by treeowl**
* Add a rule `toView (fromView m) = m >>= return`.
* Make `sconcat` total.
* Remove `SPECIALIZE INLINE` pragmas that aren't clearly useful.
* Add some documentation.

### CODE REVIEW

File: src/Control/Monad/Logic/Sequence/Internal.hs [hunk 1]
```diff
@@ -174,16 +175,27 @@ type Seq = SeqT Identity
...
+-- For now, we don't assume the monad identity law holds for the underlying
+-- monad. We may re-evaluate that later, but it's a bit tricky to document the
+-- detailed strictness requirements properly.
+--
+-- We do, however, assume that `pure /= _|_`, or that `>>=` doesn't `seq` on
+-- its second argument, and that we can therefore eta-reduce `\x -> pure x` to
+-- just `pure`. It seems quite safe to assume that at least one of these is
+-- true, since in real code they're virtually always *both* true.
+{-# RULES
+"toView . fromView" forall m. toView (fromView m) = m >>= return
```
Comment (dagit) [inline src/Control/Monad/Logic/Sequence/Internal.hs:197]: Do you see any performance difference with this? I guess what I'm actually wondering is: does this ever fire?

**Comment by dagit** [comment]
I think this is probably fine to merge. I think I got the `sconcat` from the sources for the paper. I never really thought about it being partial, but I think that change is good.

**Comment by treeowl** [comment]
I can double check that the rule sometimes fires. Ideally, we should make sure it fires pretty reliably, but that can be in a different PR.

**Comment by treeowl** [comment]
`-ddump-rule-rewrites` indicates that the rule fires 24 times while compiling the test suite. We should certainly check later to make sure we're not missing rewrite opportunities, but it at least works sometimes!

**Pull request merged**